File: tests/workflow/producer/rust/src/producer/responses/groupb_groupc_structa.rs
use super::{identification, producer::Control, protocol, scope::Scope, Context};
use crate::stat::Alias;
use crate::test::samples;
use clibri::server;

#[allow(unused_variables)]
pub async fn response<'c, E: server::Error, C: server::Control<E>>(
    request: &protocol::GroupB::GroupC::StructA,
    scope: &mut Scope<'_, E, C>,
) -> Result<protocol::GroupB::GroupC::StructB, protocol::GroupA::StructB> {
    let index = scope
        .context
        .get_index(scope.identification.uuid(), Alias::GroupBGroupCStructA);
    if index == 1 {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupBGroupCStructB);
        Ok(samples::group_b::group_c::struct_b::get())
    } else {
        scope
            .context
            .inc_stat(scope.identification.uuid(), Alias::GroupAStructB);
        Err(samples::group_a::struct_b::get())
    }
}

File: tests/workflow/producer/typescript/src/producer/responses/groupb.groupc.structa.ts
import { Response } from "../implementation/responses/groupb.groupc.structa";
import {
	Context,
	Producer,
	Identification,
	Filter,
	Protocol,
} from "../implementation/responses";
import { Alias } from "../../stat";
import { panic } from "../../tools";
import { Scope } from "../implementation/scope";

export function response(
	request: Protocol.GroupB.GroupC.StructA,
	scope: Scope
): Promise<Response> {
	const stat = scope.context.getStat(scope.consumer.uuid());
	stat.case(Alias.GroupBGroupCStructA);
	const index = stat.index(Alias.GroupBGroupCStructA);
	if (index === 0) {
		return Promise.resolve(
			new Response(Protocol.GroupA.StructB.defaults())
		);
	} else if (index === 1) {
		return Promise.resolve(
			new Response(Protocol.GroupB.GroupC.StructB.defaults())
		);
	} else {
		panic(
			`Unexpected index for Protocol.GroupB.GroupC.StructA request: ${index}`
		);
		return Promise.reject(
			new Error(
				`Handler for Protocol.GroupB.GroupC.StructA isn't implemented.`
			)
		);
	}
}
